### PR TITLE
Enable multiple SRRs as inputs in import-sra and add processes for merging FASTQs

### PR DESCRIPTION
**Pull request by jvrakor**
@jkokosar @mzganec 

## Checklist

* [x] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [x] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [x] All inputs are used in process.
* [x] All output fields have their ``re-save`` calls.

### Processes ###

* Code that was once accepted should be kept like it is. It should be changed
  only when its functionality has changed. Small parts may be changed
  to fix bugs (by changing code only to make it nicer you might introduce a bug).
* Set the optimal number of cores that the tool can use (check if parallelization
  can be performed).
* Description of tool should be self explanatory with links to tool's homepage
  and publication.
* Write explanatory input labels and add short command line parameter names
  in square brackets (for example: Maximum length [``--maxlength``]).
* Where possible provide default values for the process input fields.
  If they are provided, then ``required: false`` is not needed.
* Add descriptive error and warning messages.
* Place a ``re-checkrc`` call after each process step (call of a tool, a
  Bash command, a Python script or an R script).
* Add ``re-progress`` calls where appropriate.

### Commit and PR messages ###

* Each commit should be minimal (i.e. 1 change) and self-contained (including
  tests).
* Changes to Docker images need to be accepted in a separate pull request.
  Before the new images can be used for automated testing of your PR, they
  must be manually tagged and pushed to Docker Hub.
* Mark incomplete PRs with the [WIP] (Work In Progress) tag.
* Do not paste links to private repositories from the public one.

### Tests ###

* Add meaningful names of test files with spaces:
  * Genomes: ``genome species.fa.gz``
  * Annotations: ``annotation species.gtf.gz``
  * Adapters: ``adapters-name/type-source``.
* Output and specific input files should include the name of the tested tool.
* Keep test files' size small. If there is no other option, add the large
  test file to the ``files/large`` directory so it will be tracked with Git LFS.
* Pay attention to edge cases: what will happen if any data is
  missing, if length of some list is 0, if string is given where one expects
  integer, if file is empty, ...
* When testing workflows, try to verify workflow success by checking one
  of the last steps. Pick the one that is most likely to fail. Also test
  that all steps have 'OK' status.

### CODE REVIEW

File: resolwe_bio/processes/support_processors/merge_samples.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        with open(merged_fastq, 'wb') as outfile:
+            for reads in inputs.reads:
+                for fastq in reads.fastq:
+                    with open(fastq.path, 'rb') as infile:
+                        for line in infile:
+                            outfile.write(line)
+
+        outputs.fastq = [merged_fastq]
+
+        print('Postprocessing FastQC...')
```
Comment (jkokosar) [inline resolwe_bio/processes/support_processors/merge_samples.py]: FastQC operation is done twice in two processes within the same file. This could be generalised into function defined in this file. However, it is not yet possible to import this function from other location to be used in other processes too.

**Comment by jkokosar** [comment]
Thank you for taking the time to port the processes to Python. Thats really nice. The general concept is good!

File: resolwe_bio/processes/import_data/sra_file.py [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+        # FastQC writes both progress and errors to stderr and exits with code 0.
+        # Catch if file is empty, wrong format... (Failed to process) or
+        # if file path does not exist, file cannot be read (Skipping).
+        if 'Failed to process' in stderr or 'Skipping' in stderr:
+            self.error('Failed while processing with FastQC.')
+
+        for fastqc_zip in glob.glob('fastqc/*_fastqc.zip'):
+            move(fastqc_zip, '.')
+
+        for report_dir in next(os.walk('fastqc'))[1]:
```
[the remark this answers is not in the document]
Comment (jvrakor) [reply resolwe_bio/processes/import_data/sra_file.py]: I was expecting this one :)

File: resolwe_bio/processes/import_data/sra_file.py [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+                cmd = cmd['--clip']
+            if inputs.advanced.aligned:
+                cmd = cmd['--aligned']
+            if inputs.advanced.unaligned:
+                cmd = cmd['--unaligned']
+            return_code, _, _ = cmd[srr] & TEE(retcode=None)
+            if return_code:
+                self.error(f'Download of {srr} reads with fastq-dump failed.')
+
+        fastq_gz = glob.glob('*.fastq.gz')
```
Comment (romunov) [inline resolwe_bio/processes/import_data/sra_file.py]: You can use `pathlib.glob` function to gather files with a specific pattern. There are a number of occurrences throughout the document.

File: resolwe_bio/processes/import_data/sra_file.py [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+        outputs.fastq = fastq_gz
+
+        print('Postprocessing FastQC...')
```
Comment (romunov) [inline resolwe_bio/processes/import_data/sra_file.py]: Consider using `self.info()`.

File: resolwe_bio/processes/import_data/sra_file.py [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+            move(fastqc_zip, '.')
+
+        fastqc_url = []
+        fastqc_dir = Path('fastqc')
+        for report_dir in fastqc_dir.iterdir():
+            if not report_dir.is_dir():
+                continue
+
+            fastqc_url.append({
+                "file": str(report_dir / 'fastqc_report.html'),
```
Comment (romunov) [inline resolwe_bio/processes/import_data/sra_file.py]: Instead of `str()` you could use `x.name` which could be prettier to read. Maybe.

File: resolwe_bio/processes/import_data/sra_file.py [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+                "refs": [str(report_dir)],
+            })
+
+            reads_name = report_dir.name.replace('_fastqc', '')
+            encoding_file = report_dir / 'fastqc_data.txt'
+            encoding = Cmd['parse_encoding_type.py'](str(encoding_file))
+
+            if encoding == 'Illumina 1.5' or encoding == 'Illumina 1.3':
+                print("Recoding input reads from Phred64 encoding to Phred33 encoding.")
```
Comment (romunov) [inline resolwe_bio/processes/import_data/sra_file.py]: See comment on `self.info()` above.

File: resolwe_bio/processes/import_data/sra_file.py [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+                os.rename(f'{reads_name}.fastq.gz', 'input_reads.fastq.gz')
```
Comment (romunov) [inline resolwe_bio/processes/import_data/sra_file.py]: `pathlib` has a [`rename` method](https://docs.python.org/3/library/pathlib.html#pathlib.Path.rename).

File: resolwe_bio/processes/import_data/sra_file.py [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+                cmd = Cmd[
+                    'TrimmomaticSE',
+                    '-phred64',
+                    'input_reads.fastq.gz',
+                    'reformated.fastq.gz',
+                    'TOPHRED33',
+                ]()
+                os.rename('reformated.fastq.gz', f'{reads_name}.fastq.gz')
```
Comment (romunov) [inline resolwe_bio/processes/import_data/sra_file.py]: See comment about `pathlib.rename` method above.

File: resolwe_bio/processes/import_data/sra_file.py [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+        sra_accession = ListField(StringField(), label='SRA accession(s)')
+        show_advanced = BooleanField(label='Show advanced options', default=False)
+
+        class Advanced:
+            """Advanced options."""
+            min_spot_id = IntegerField(label='Minimum spot ID', required=False)
+            max_spot_id = IntegerField(label='Maximum spot ID', required=False)
+            min_read_len = IntegerField(label='Minimum read length', required=False)
+            clip = BooleanField(label='Clip adapter sequences', default=False)
+            aligned = BooleanField(label='Dump only aligned sequences', default=False)
```
Comment (romunov) [inline resolwe_bio/processes/import_data/sra_file.py]: Are these two options (`aligned`, `unaligned`) mutually exclusive? If yes, can we make them like that here?

File: resolwe_bio/processes/import_data/sra_file.py [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+        fastq2 = ListField(FileField(), label='Reads file (mate 2)')
+        fastqc_url = ListField(
+            FileHtmlField(),
+            label='Quality control with FastQC (Upstream)',
+        )
+        fastqc_url2 = ListField(
+            FileHtmlField(),
+            label='Quality control with FastQC (Downstream)',
+        )
+        fastqc_archive = ListField(FileField(), label='Download FastQC archive (Upstream)')
```
Comment (romunov) [inline resolwe_bio/processes/import_data/sra_file.py]: Do we want to use consistent mate1/mate2 annotation here in place of upstream/downstream?

File: resolwe_bio/processes/import_data/sra_file.py [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+                continue
+
+            reads_name = report_dir.name.replace('_fastqc', '')
+            encoding_file = report_dir / 'fastqc_data.txt'
+            encoding = Cmd['parse_encoding_type.py'](str(encoding_file))
+
+            if encoding == 'Illumina 1.5' or encoding == 'Illumina 1.3':
+                print("Recoding input reads from Phred64 encoding to Phred33 encoding.")
+                os.rename(f'{reads_name}.fastq.gz', 'input_reads.fastq.gz')
+                cmd = Cmd[
```
Comment (romunov) [inline resolwe_bio/processes/import_data/sra_file.py]: `cmd` value not used. Perhaps skip assigning to a variable?

File: resolwe_bio/processes/import_data/sra_file.py [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+                    'TrimmomaticSE',
+                    '-phred64',
+                    'input_reads.fastq.gz',
+                    'reformated.fastq.gz',
+                    'TOPHRED33',
+                ]()
+                os.rename('reformated.fastq.gz', f'{reads_name}.fastq.gz')
```
Comment (romunov) [inline resolwe_bio/processes/import_data/sra_file.py]: See comment about `pathlib.rename` method above.

File: resolwe_bio/processes/support_processors/merge_samples.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
```
[the remark this answers is not in the document]
Comment (jkokosar) [reply resolwe_bio/processes/support_processors/merge_samples.py]: This will print to the process state (info, warning, error) which is not the goal here. This print statement is there to have a better overview over the stdout output.

File: resolwe_bio/processes/import_data/sra_file.py (hunk 10 above)
Comment (jkokosar) [reply resolwe_bio/processes/import_data/sra_file.py]: mate1/mate2 is the preferred way, yes!

File: resolwe_bio/processes/support_processors/merge_samples.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+        fastq = ListField(FileField(), label='Reads file')
+        fastqc_url = ListField(
+            FileHtmlField(),
+            label='Quality control with FastQC',
+        )
+        fastqc_archive = ListField(FileField(), label='Download FastQC archive')
+
+    def run(self, inputs, outputs):
+        """Run the analysis."""
+        fastqc = []
```
[the remark this answers is not in the document]
Comment (jvrakor) [reply resolwe_bio/processes/support_processors/merge_samples.py]: good catch

File: resolwe_bio/processes/import_data/sra_file.py (hunk 9 above)
Comment (jvrakor) [reply resolwe_bio/processes/import_data/sra_file.py]: @mzganec

File: resolwe_bio/processes/support_processors/merge_samples.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+        }]
+        report_dir_2 = fastqc_path / f'{name_2}_fastqc'
+        fastqc_url_2 = [{
+            "file": str(report_dir_2 / "fastqc_report.html"),
+            "refs": [str(report_dir_2)],
+        }]
+
+        print('Postprocessing FastQC...')
+        stderr = run_fastqc([merged_fastq_1, merged_fastq_2], './fastqc')
+        # FastQC writes both progress and errors to stderr and exits with code 0.
```
[the remark this answers is not in the document]
Comment (jvrakor) [reply resolwe_bio/processes/support_processors/merge_samples.py]: Definitely

File: resolwe_bio/processes/import_data/sra_file.py [hunk 16]
```diff
@@ -0,0 +1,403 @@
...
+
+    :param list fastqs: List of fastqs
+    :param str output_dir: Output directory
+
+    """
+    output_path = Path(output_dir)
+    if output_path.exists() and not output_path.is_dir():
+        raise TypeError(f"'{output_dir}' already exists but is not a directory.")
+    elif not output_path.exists():
+        os.mkdir(str(output_path))
```
[the remark this answers is not in the document]
Comment (jvrakor) [reply resolwe_bio/processes/import_data/sra_file.py]: nice

File: resolwe_bio/processes/import_data/sra_file.py (hunk 5 above)
Comment (jvrakor) [reply resolwe_bio/processes/import_data/sra_file.py]: Prettier to read but we want full paths and not only filenames here.

File: resolwe_bio/processes/import_data/sra_file.py [hunk 17]
```diff
@@ -0,0 +1,403 @@
...
+    cmd = cmd[f'--outdir={str(output_path)}']
+    _, _, stderr = cmd & TEE
+
+    return stderr
+
+
+class ImportSra(Process):
+    """Import reads from SRA.
+
+    Import single or paired-end reads from Sequence Read Archive (SRA) via an SRA
```
[the remark this answers is not in the document]
Comment (jvrakor) [reply resolwe_bio/processes/import_data/sra_file.py]: I like that.

File: resolwe_bio/processes/import_data/sra_file.py (hunk 7 above)
Comment (jvrakor) [reply resolwe_bio/processes/import_data/sra_file.py]: Also nice.

File: resolwe_bio/processes/chip_seq/macs2.yml [hunk 18]
```diff
@@ -863,7 +863,7 @@
             awk 'BEGIN{OFS="\t"}{$4="N";$5="1000";print $0}' > "${TAGALIGN}"
             re-checkrc "Conversion to tagAlign failed"
           else
-            samtools sort -n {{ case.bam.file }} "${CASE_NAME}_sorted"
+            samtools sort -n {{ case.bam.file }} -o "${CASE_NAME}_sorted.bam"
```
Comment (jkokosar) [inline resolwe_bio/processes/chip_seq/macs2.yml]: Bump the process version

File: docs/CHANGELOG.rst [hunk 19]
```diff
@@ -9,6 +9,12 @@ This project adheres to `Semantic Versioning <http://semver.org/>`_.
 Unreleased
 ==========
 
+Add
+---
+- Add ``merge-fastq-single`` and ``merge-fastq-single`` processes that
```
Comment (jkokosar) [inline docs/CHANGELOG.rst]: Add ``merge-fastq-single`` and ``merge-fastq-paired`` ...

File: resolwe_bio/processes/import_data/sra_file.py [hunk 20]
```diff
@@ -0,0 +1,399 @@
...
+            min_read_len = IntegerField(label="Minimum read length", required=False)
+            clip = BooleanField(label="Clip adapter sequences", default=False)
+            aligned = BooleanField(label="Dump only aligned sequences", default=False)
+            unaligned = BooleanField(label="Dump only unaligned sequences", default=False)
+
+        advanced = GroupField(Advanced, label="Advanced options", hidden="!show_advanced")
+
+    def run(self, inputs, outputs):
+        """Run the analysis."""
+        first_read = Cmd["fastq-dump"]("-X", 1, "-Z", "--split-spot", inputs.sra_accession[0])
```
Comment (mzganec) [inline resolwe_bio/processes/import_data/sra_file.py]: This test could be more thorough, failing for SRAs of mixed types.

Comment (jvrakor) [reply resolwe_bio/processes/import_data/sra_file.py]: Agreed, I will add that.
